File: api/src/main/java/com/example/demo/model/ProjetoFaseTarefa.java
package com.example.demo.model;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import lombok.Data;

import javax.persistence.*;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

@Data
@Entity
@Table(name = "projeto_fase_tarefa")
public class ProjetoFaseTarefa {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "codigo")
    private String codigo;

    @ManyToOne
    @JoinColumn(name = "projeto_fase_id")
    private ProjetoFase projetoFase;

    @Column(name = "andamento")
    private Double andamento;

    @ManyToOne
    @JoinColumn(name = "tarefa_id")
    private Tarefa tarefa;

    @ManyToOne
    @JoinColumn(name = "usuario_responsavel_id")
    private Usuario usuarioResponsavel;

    @Column(name = "descricao")
    private String descricao;

    @Column(name = "data_inicio")
    private LocalDate dataInicio;

    @Column(name = "data_fim")
    private LocalDate dataFim;

    @ManyToOne
    @JoinColumn(name = "projeto_fase_tarefa_dependente_id")
    @JsonIgnoreProperties(value = {"recursos", "tarefa", "usuarioResponsavel", "projetoFase", "projetoFaseTarefaDependente"})
    private ProjetoFaseTarefa projetoFaseTarefaDependente;

    @ElementCollection
    @CollectionTable(name = "projeto_fase_tarefa_recurso", joinColumns = {@JoinColumn(name = "projeto_fase_tarefa_id")})
    private List<ProjetoFaseTarefaRecurso> recursos = new ArrayList<>();

}

File: ui/src/app/model/projeto-fase-tarefa.ts
import { FormControl, FormGroup, Validators, FormArray } from '@angular/forms';

import { Tarefa } from './tarefa';
import { ProjetoFase } from './projeto-fase';
import { Usuario } from './usuario';
import { ProjetoFaseTarefaRecurso } from './projeto-fase-tarefa-recurso';

export class ProjetoFaseTarefa {

  id: number | undefined;
  codigo: string | undefined;
  projetoFase: ProjetoFase | undefined;
  tarefa: Tarefa | undefined;
  usuarioResponsavel: Usuario | undefined;
  descricao: string | undefined;
  andamento: number | undefined;
  dataInicio: Date | undefined;
  dataFim: Date | undefined;
  projetoFaseTarefaDependente: ProjetoFaseTarefa | undefined;
  recursos: ProjetoFaseTarefaRecurso [] | undefined;

  constructor() {
    this.recursos = [];
    this.andamento = 0;
   }

   criarFormulario(projetoFaseTarefa: ProjetoFaseTarefa) {
    return new FormGroup({
      id: new FormControl(projetoFaseTarefa.id),
      codigo: new FormControl(projetoFaseTarefa.codigo, Validators.required),
      projetoFase: new FormControl(projetoFaseTarefa.projetoFase, Validators.required),
      tarefa: new FormControl(projetoFaseTarefa.tarefa, Validators.required),
      usuarioResponsavel: new FormControl(projetoFaseTarefa.usuarioResponsavel, Validators.required),
      descricao: new FormControl(projetoFaseTarefa.descricao, Validators.required),
      andamento: new FormControl(projetoFaseTarefa.andamento),
      dataInicio: new FormControl(projetoFaseTarefa.dataInicio, Validators.required),
      dataFim: new FormControl(projetoFaseTarefa.dataFim, Validators.required),
      projetoFaseTarefaDependente: new FormControl(projetoFaseTarefa.projetoFaseTarefaDependente),
      recursos: new FormArray(projetoFaseTarefa!.recursos!.map(recurso => ProjetoFaseTarefaRecurso.criarFormulario(recurso)))
    });
  }

}
